hail/src/main/scala/is/hail/expr/ir/Interpret.scala

```diff
@@ -702,8 +702,25 @@ object Interpret {
             FastIndexedSeq(classInfo[Region], LongInfo, LongInfo), UnitInfo,
             extracted.seqPerElt)
 
-          val read = extracted.deserialize(ctx, spec, physicalAggs)
-          val write = extracted.serialize(ctx, spec, physicalAggs)
+          case class WrappedByteArray(var bytes: Array[Byte])
+
+          val read = {
```

Comment:
That's a correct summary, though I prefer to reason about the correctness of components in isolation, rather than tracing the global path of each region.

The key concept is ownership. Each region holding an aggregator state in the combiner code has a unique owner at all times. Technically, an owner is a reference, which might be a local variable in a stack frame, or a reference inside some data structure like the AssociativeCombiner, but it's easier to think of the owner being the function or data structure itself.

Functions that create regions like `Region.apply` or `read` here give ownership of the new region to their caller. Functions that consume regions like `Region.invalidate` or `write` here take ownership from their caller. `clear` does not take ownership, since it's really an optimized form of `r.invalidate(); r = Region()`.

The rules for a function are:
* A function must not own any regions at the end of its body, i.e. for each region of which ownership is given, ownership must be given away.
* If ownership of a region is given away, that region may not be referenced by the function again. We have to assume the callee freed the region.
* If the function is given a non-owning reference to a region, we must assume the region is freed immediately after the function returns. In particular, we should not store a reference anywhere that will live longer than the function call.

Similarly for a data structure, but instead of reasoning over the duration of the function call, we have to think about the lifetime of the data structure, which ideally corresponds to the duration of an owning function.

So for example, (the function returned by) `combOpF` is correct because it is given ownership of two regions, one it gives to `invalidate`, the other it returns back to the caller. No need to think about entire lifetime of any one region.

The correctness of `AssociativeCombiner` can be reasoned about simply using the invariant: every region in the TreeMap is owned by the combiner. Then you can walk through the `combine` method code and see that every region it is given ownership of (the input value, values taken from the TreeMap, or values returned by the combOp) is given to someone else (passed to combOp or put into the TreeMap). And when `result` is called, we assert that the TreeMap has 0 or 1 element; if 1 we return it to the caller, so we're left owning zero regions, and if 0 we create a new region holding the zero value and return that.

The local rules ensure that every region has a unique owner at any point in the execution. Therefore, if at the beginning of a block of code, like the `TableAggregate` interpret case, there are no owned regions, and at the end there are no owned regions, then we know that every region that was created was freed, without needing to know about the implementation of any methods we called, other than their ownership signature.

Reply:
Thanks! I think my problem here is that since the `rvd.combine` function we're passing all of these pieces into doesn't think about regions, I need to follow that chain of function (into `rvd.combine` and the combiner stuff) in order to reason about the flow of region values through the various functions. Adding comments here is helpful, because at least now, when I do that, I already know what each function is supposed to do w.r.t region ownership when I look at that code.

Reply:
Yeah, you're right. `rvd.combine` doesn't need to know about regions to be correct, but it does need to know that the `U` type might contain some kind of resource, and that the `combOp`, `read`, and `write` functions have the corresponding ownership semantics for that resource. I should really have added comments there too.